src/dict.c

```diff
@@ -1072,10 +1077,12 @@ static int _dictExpandIfNeeded(dict *d)
      * table (global setting) or we should avoid it but the ratio between
      * elements/buckets is over the "safe" threshold, we resize doubling
      * the number of buckets. */
-    if (d->ht_used[0] >= DICTHT_SIZE(d->ht_size_exp[0]) &&
-        (dict_can_resize ||
-         d->ht_used[0]/ DICTHT_SIZE(d->ht_size_exp[0]) > dict_force_resize_ratio) &&
-        dictTypeExpandAllowed(d))
+    if (!dictTypeExpandAllowed(d))
+        return DICT_OK;
+    if ((dict_can_resize == DICT_RESIZE_ENABLE &&
+         d->ht_used[0] >= DICTHT_SIZE(d->ht_size_exp[0])) ||
+        (dict_can_resize != DICT_RESIZE_FORBID &&
+         d->ht_used[0] / DICTHT_SIZE(d->ht_size_exp[0]) > dict_force_resize_ratio))
```

Comment:
it seem better to check ratio first and then dictTypeExpandAllowed

The new code:
```c
    if ((dict_can_resize == DICT_RESIZE_ENABLE &&
         d->ht_used[0] >= DICTHT_SIZE(d->ht_size_exp[0])) ||
        (dict_can_resize != DICT_RESIZE_FORBID &&
         d->ht_used[0] / DICTHT_SIZE(d->ht_size_exp[0]) > dict_force_resize_ratio))
    {
        if (!dictTypeExpandAllowed(d))
            return DICT_OK;
        return dictExpand(d, d->ht_used[0] + 1);
    }
```

I noticed that if this is changed to new code, set will have a 5% performance improvement.

There are 11032768 fixed keys in the database
`redis-server --maxmemory big_number --save ""`
`src/redis-benchmark -P 100 -r 1000000000 -t set -n 5000000`
`throughput summary: n requests per second`

unstable:
```
848032.56
854408.69
858663.94
871839.56
882612.56
```

new code:
```
897988.56
913408.88
914076.81
916758.31
920640.75
```

do you think we should take a shot?

Reply:
What I'm wondering is whether there might be performance degradation when `expandAllowed` is not null.

Reply:
The suggestion makes sense to me. `dictExpandAllowed` (for the main db dict and the expire dict) seems to involve a few function calls and memory accesses, so let's do it only after the `>=` and `>` checks. (We can swap the division for a multiplication too on the right hand side, if it makes a difference.)

When expandAllowed is null, then expand is always allowed so it shouldn't matter if the check is done first or last.

Reply:
so you're saying that this PR added a 5% performance. i think we should have noticed. @filipecosta90 FYI.
note that we backported this fix to several releases.

the suggested change seems good to me.
but i suppose we'll still hit the degradation when over the thresholds, so i wonder if we should optimize `overMaxmemoryAfterAlloc` in some way? like caching the memory metrics, and only refreshing them in beforeSleep or serverCron?

Reply:
yes, please make a PR, we can explore additional changes there (i'm afraid to forget it when the discussion is in an old PR).